### DISCUSSION TITLE
How to use ignore file from a child directory

### DISCUSSION THREAD

**Discussion by mhanberg** [BurntSushi/ripgrep#1977] [Q&A] [answered]
Hi!

I am trying to write an alias to quickly search for a string in all projects.

What i'm noticing is that things like dependency folders (like `deps` in Elixir or `node_modules` in npm` are showing up, so I assume that the .gitignore file is not being respected.

Basically, I'd like to run ripgrep from directory A, but respect the gitignore files for each project `P1`, `P2`, etc

In this example, assume that the git ignore in P1 has the normal npm ignore globs and P2 has the normal Elixir ignore globs. I would want to be able to run ripgrep in directory A, but for it to ignore the node_modules files and the deps files

```
A/P1/.gitignore
A/P1/node_modules/foo/index.js
A/P1/src/bar.js
A/P1/src/baz.js
A/P1/src/bob.js
A/P2/.gitignore
A/P2/deps/foo/main.ex
A/P2/lib/bar.ex
A/P2/lib/baz.ex
A/P2/lib/bob.ex
```

Is this possible?

**Comment by BurntSushi** [comment, marked as answer]
Yes. It should work. Consider this directory tree:

```
$ tree -a
.
├── p1
│   ├── foo.rs
│   ├── .gitignore
│   └── node_modules
│       └── foo.rs
└── p2
    ├── deps
    │   └── foo.rs
    ├── foo.rs
    └── .gitignore
```

With the contents of `.gitignore`:

```
$ cat p1/.gitignore
/node_modules
$ cat p2/.gitignore
/deps
```

Let's run `rg --files` to show what will be searched:

```
$ rg --files
p2/foo.rs
p1/foo.rs
p1/node_modules/foo.rs
p2/deps/foo.rs
```

But wait, why are the `node_modules` and `deps` files showing up? That's because neither `p1` nor `p2` are actually git repos. ripgrep only respects gitignore files inside git repos, and ripgrep detects git repos by the presence or absence of `.git`. So let's create that:

```
$ mkdir {p1,p2}/.git
```

And now, `rg --files` shows what you would expect:

```
$ rg --files
p2/foo.rs
p1/foo.rs
```

So if this isn't working for you, then I think you'll need to give me something that I can reproduce.

**Reply by mhanberg** [reply]
Thanks for the explanation, I'll make sure I'm not doing anything else wrong first, and if I still can't figure it out i'll bring a repo case.

Thanks for the response!

**Reply by BurntSushi** [reply]
Aye. Don't forget about the `--debug` flag. It should tell you which file is being ignored and why.

**Reply by mhanberg** [reply]
ahh, I figured it out. Your bit about only using .gitignore for actual git repos was the missing piece.

I had one project that was just plain strange as well as 3 that weren't git repos... 😅.

I appreciate your time and help here, thanks!

**Reply by BurntSushi** [reply]
Yeah I figured that was the case! Glad it's working!
